libsplinter/src/admin/store/mod.rs

```diff
@@ -0,0 +1,454 @@
+                &proposal.circuit.circuit_management_type == man_type
+            }
+            CircuitPredicate::Members(nodes) => {
+                for node_id in nodes {
+                    if proposal
+                        .circuit
+                        .members
+                        .iter()
+                        .find(|node| node_id == &node.node_id)
+                        .is_none()
+                    {
+                        return false;
+                    }
+                }
+                true
+            }
+        }
+    }
+}
+
+/// Defines methods for CRUD operations and fetching and listing circuits, proposals, nodes and
+/// services without defining a storage strategy
+pub trait AdminServiceStore: Send + Sync {
+    /// Adds a circuit proposal to the underlying storage
+    ///
+    /// # Arguments
+    ///
+    ///  * `proposal` - The proposal to be added
+    ///
+    ///  Returns an error if a `CircuitProposal` with the same ID already exists
+    fn add_proposal(&self, proposal: CircuitProposal) -> Result<(), AdminServiceStoreError>;
+
+    /// Updates a circuit proposal in the underlying storage
+    ///
+    /// # Arguments
+    ///
+    ///  * `proposal` - The proposal with the updated information
+    ///
+    ///  Returns an error if a `CircuitProposal` with the same ID does not exist
+    fn update_proposal(&self, proposal: CircuitProposal) -> Result<(), AdminServiceStoreError>;
+
+    /// Removes a circuit proposal from the underlying storage
+    ///
+    /// # Arguments
+    ///
+    ///  * `proposal_id` - The unique ID of the circuit proposal to be removed
+    ///
+    ///  Returns an error if a `CircuitProposal` with specified ID does not exist
+    fn remove_proposal(&self, proposal_id: &str) -> Result<(), AdminServiceStoreError>;
+
+    /// Fetches a circuit proposal from the underlying storage
+    ///
+    /// # Arguments
+    ///
+    ///  * `proposal_id` - The unique ID of the circuit proposal to be returned
+    fn fetch_proposal(
+        &self,
+        proposal_id: &str,
+    ) -> Result<Option<CircuitProposal>, AdminServiceStoreError>;
+
+    /// List circuit proposals from the underlying storage
+    ///
+    /// The proposals returned can be filtered by provided CircuitPredicate. This enables
+    /// filtering by management type and members.
+    fn list_proposals(
+        &self,
+        predicates: &[CircuitPredicate],
+    ) -> Result<Box<dyn ExactSizeIterator<Item = CircuitProposal>>, AdminServiceStoreError>;
+
+    /// Adds a circuit to the underlying storage. Also includes the associated Services and
+    /// Nodes
+    ///
+    /// # Arguments
+    ///
+    ///  * `circuit` - The circuit to be added to state
+    ///  * `nodes` - A list of nodes that represent the circuit's members
+    ///
+    ///  Returns an error if a `Circuit` with the same ID already exists
+    fn add_circuit(
+        &self,
+        circuit: Circuit,
+        nodes: Vec<CircuitNode>,
```

Comment:
No, nodes are stored separately because they can be referenced by multiple circuits. They are included in the circuit defined in a proposal but not in the actual circuit definition.